src/lib.rs

```diff
@@ -1,95 +1,112 @@
-extern crate ignore;
-extern crate regex;
-use std::{cmp, sync::mpsc::{self, Sender}};
+#![warn(missing_docs)]
+// Use the readme as the crate documentation
+#![doc = include_str!("../README.md")]
+
+use std::{
+    cmp,
+    path::Path,
+    sync::mpsc::{self, Sender},
+};
+
 use ignore::{WalkBuilder, WalkState};
-use regex::{Regex, RegexBuilder};
+use regex::Regex;
 
-pub enum Depth {
-    None,
-    Some(u8),
+/// A struct that holds the receiver for the search results
+///
+/// Can be iterated on to get the next element in the search results
+///
+/// # Examples
+///
+/// ## Iterate on the results
+///
+/// ```
+/// use rust_search::Search;
+///
+/// let search = Search::new("src", None, Some(".rs"), Some(1));
+///
+/// for path in search {
+///    println!("{:?}", path);
+/// }
+/// ```
+///
+/// ## Collect results into a vector
+///
+/// ```
+/// use rust_search::Search;
+///
+/// let search = Search::new("src", None, Some(".rs"), Some(1));
+///
+/// let paths_vec: Vec<String> = search.collect();
+/// ```
+pub struct Search {
+    rx: mpsc::Receiver<String>,
 }
 
-pub enum FileType<'a> {
-    None,
-    Some(&'a str),
-}
+impl Iterator for Search {
+    type Item = String;
 
-pub enum SearchInput<'a> {
-    None,
-    Some(&'a str),
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.rx.recv() {
+            Ok(v) => Some(v),
+            Err(_) => None,
+        }
+    }
 }
 
-pub fn get_paths (search_location: &str, search_input: SearchInput, file_type: FileType, depth: Depth) -> std::sync::mpsc::Receiver<String>{
-    let regex_search_input = build_regex_search_input(search_input, file_type);
-    
-    let depth = match depth {
-        Depth::None => None,
-        Depth::Some(depth) => Some(depth as usize),
-    };
-
-    let walker = WalkBuilder::new(search_location)
-        .hidden(true)
-        .git_ignore(true)
-        .max_depth(depth)
-        .threads(threads())
-        .build_parallel();
-
-
-    let (tx, rx) = mpsc::channel::<String>();
-
-    walker.run(|| {
-        let tx: Sender<String> = tx.clone();
-        let reg_exp: Regex = regex_search_input.clone();
-        let exclude_directories: bool = false;
-
-        Box::new(move |path_entry: Result<ignore::DirEntry, ignore::Error>| {
-            if let Ok(entry) = path_entry {
-                if exclude_directories && !entry.path().is_file() {
-                    WalkState::Continue
-                } else {
+impl Search {
+    /// Create a new instance of the searcher
+    pub fn new(
+        search_location: impl AsRef<Path>,
+        search_input: Option<&str>,
+        file_type: Option<&str>,
+        depth: Option<usize>,
+    ) -> Self {
+        let regex_search_input = build_regex_search_input(search_input, file_type);
+
+        let walker = WalkBuilder::new(search_location)
+            .hidden(true)
+            .git_ignore(true)
+            .max_depth(depth)
+            .threads(cmp::min(12, num_cpus::get()))
```

Comment:
Does this mean we are getting the minimum of both? And why is that?

Reply:
That will find the minimum of the two numbers, and I am honestly not sure that was already in the code when I forked it.